### PR TITLE
Add functions for matrix multiplication

### PR DESCRIPTION
**Pull request by fangxiaoran**

### CODE REVIEW

**Review by justalittlenoob — APPROVED** [review]
Looks great.

File: src/ipcl_python/ipcl_python.py [hunk 1]
```diff
@@ -348,42 +345,38 @@ def apply_obfuscator(self):
...
+                raise IndexError("__getitem__: key out of range")
+            return PaillierEncryptedNumber(
+                self.public_key,
+                self.__ipclCipherText.getCipherText(key),
+                [self.exponent(key)],
+                1,
+            )
+
+    def __setitem__(self, *key):
+        print("__setitem__: Not supported")
```
Comment (hamishun) [inline src/ipcl_python/ipcl_python.py]: If not supported, why keep the method?

File: src/ipcl_python/ipcl_python.py [hunk 2]
```diff
@@ -826,13 +828,237 @@ def mean(self) -> "PaillierEncryptedNumber":
...
+            raise ValueError(
+                "PaillierEncryptedNumber.matmul: matrix multiply size mismatch"
+            )
+
+        other = np.array(other)
+        if other.ndim == 1 or other.ndim == 2:
+            # self.shape: (m x n), other.shape: (n x k), k could be none
+            n = other.shape[0]
+            k = other.shape[1] if other.ndim == 2 else 1
+            m = self.__len__() // n
```
Comment (hamishun) [inline src/ipcl_python/ipcl_python.py]:
```suggestion
            m = len(self) // n
```

File: src/ipcl_python/ipcl_python.py [hunk 3]
```diff
@@ -348,42 +345,38 @@ def apply_obfuscator(self):
...
+                [self.exponent(key)],
+                1,
+            )
+
+    def __setitem__(self, *key):
+        print("__setitem__: Not supported")
+
+    def __iter__(self) -> "PaillierEncryptedNumber":
+        for i in range(len(self)):
+            yield self.__getitem__(i)
```
Comment (hamishun) [inline src/ipcl_python/ipcl_python.py:379]: Should I think do the same as you return a simple generator object.
```suggestion
        return (self[i] for i, _ in enumerate(self))
```

File: src/ipcl_python/ipcl_python.py [hunk 4]
```diff
@@ -421,7 +414,9 @@ def __rsub__(
         self,
         other: Union["PaillierEncryptedNumber", np.ndarray, list, int, float],
     ) -> "PaillierEncryptedNumber":
-        return self.__sub__(other)
+        if isinstance(other, PaillierEncryptedNumber):
+            return other.__sub__(self)
+        return self.__mul__(-1.0).__raw_add(other)
```
Comment (hamishun) [inline src/ipcl_python/ipcl_python.py]: You shouldn't need to call the dunder methods in normal syntax, but use the symbols and built-in functions instead. That is the advantage of defining the dunder methods (a.k.a. operator overloading :))

**Review by hamishun — COMMENTED** [review]
Not to block the merge, but a few comments in case you wish to refactor.

File: src/ipcl_python/ipcl_python.py (hunk 3 above)
Comment (hamishun) [inline src/ipcl_python/ipcl_python.py]: This should work the same by returning a generator object.
```suggestion
        return (self[i] for i in range(len(self)))
```

**Comment by fangxiaoran** [comment]
> Not to block the merge, but a few comments in case you wish to refactor.

Thanks Hamish for the comments. It's really helpful! I will update the code accordingly.

**Pull request merged**